Add embedding service for issue 111

hanna-paasivirta opened the pull request:
## Short Description

Create an embedding service to access/create a embedding database and search it with an input text.

Fixes #111 

## Implementation Details
This module is primarily built for the Vocabulary Mapping project #109, but it can be used in any future projects that require embeddings. It leverages the LangChain library to allow for easy access to different types of embeddings and embedding storage services.

The main functions are:
- `create_vectorstore()` -> Create a new embedding database collection with texts 
- `get_existing_vectorstore()` -> Access an existing embedding database collection
- `get_similar_texts()` -> Use an input text to query the embedding collection for similar texts

Currently the module only allows Zilliz as a vector store and OpenAI Embeddings as a model (both require credentials). New options can be easily enabled.

## AI Usage

Please disclose how you've used AI in this work (it's cool, we just want to know!):

- [x] Code generation (copilot but not intellisense)
- [x] Learning or fact checking
- [x] Strategy / design
- [x] Optimisation / refactoring
- [x] Translation / spellchecking / doc gen
- [ ] Other
- [ ] I have not used AI

You can read more details in our [Responsible AI Policy](https://www.openfn.org/ai#pull-request-templates)

josephjclark commented:
Thank you @hanna-paasivirta ! This is exciting. I'll take a close look and see if I have any suggestions, but I'm keen for this to be a rough first cut which we'll iterate on over the coming weeks.

We'll need to break up #111 into some smaller issues to represent next steps (eg: port the docs importer into this embedding service, investigate and setup PG vector, etc)

josephjclark commented on services/embeddings/generate_embeddings.py:
```diff
@@ -0,0 +1,182 @@
+    Retrieve similar texts from a vectorstore based on input text.
+    
+    Args:
+        input_text: Text string to use for similarity search
```
Ah I'm interested in this - so we don't need to manually generate embeddings for our search strings? Langchain does this for us based on the built-in embedding function?

This is an area I've been curious about.

hanna-paasivirta replied:
No we don't, although we could if we wanted to. It'll just use the embedding type associated with the vector store.

josephjclark reviewed (commented):
This is great Hanna thanks.

I've left a couple of notes on comments and general organisation.

Let's have a chat in the morning about the main interface and how this service will actually be used. Subject to that conversation I might have more suggestions or we'll merge in and move forward :muscle:

josephjclark commented on services/embeddings/demo_data/data.json:
```diff
@@ -0,0 +1,59 @@
+"timestamp_ms": 1057190300000
+},
+{
+"content": "I must confess, the OpenFn integrations have transformed our estate management beyond measure.",
```
I love this.

josephjclark commented on services/embeddings/demo_data/data.json:
```diff
@@ -0,0 +1,59 @@
+"timestamp_ms": 1057188000000
+},
+{
+"content": "Indeed, Mr. Darcy. May your workflow automations bring you much efficiency.",
```
But what happens next??

josephjclark commented on services/embeddings/generate_embeddings.py:
```diff
@@ -0,0 +1,138 @@
+    def _get_vectorstore_class(self):
+        """Get vectorstore class based on the specified type."""
+        try:
+            return self.vectorstore_classes[self.vectorstore_type]
```
Out of curiosity, are these vectorstore classes expensive to instantiate?

Do they try to maintain a live collection to the DB or are they "lazy" and just call out when needed?

josephjclark replied:
@hanna-paasivirta This question is still outstanding :) It affects how and when we initialise our vector stores.

hanna-paasivirta replied:
This function just gets the LangChain class for a vectorstore. But in general, I'm not sure how to answer. I think the cost of instantiating a vectorstore in the code is negligible. The costly operations are generating embeddings, indexing, storing, similarity search, and insertion.

josephjclark reviewed (changes requested):
Hey @hanna-paasivirta 

I've left a few comments inline but the headline here is that the structure of this isn't what I was expecting after where we ended up in our chat yesterday.

The value of this branch is in setting up a _pattern_ that makes it easy for apollo users to create and search embedded datastores. That pattern will evolve as we move forward, but I want to get us off to a clean start. And more importantly, make sure we're on the same page and moving in the same direction.

To better explain my expectations, I've raised a draft PR #121 - please have a careful read through that and let me know what you think.

One other thing for when the code is done:  can I ask you please to create a simple, very high-level readme file which broadly explains what the embedding service is (or aiming to be), and how datastores are expeted to be used (we talked yesterday about the `stores/*`  folder structure and some kind of import of a search function). I'm looking for something really short, without much detail because things are likely to change quickly. Just an overview in maybe 100 words to give context to, for example, Elias if he were to look at this folder tomorrow and wonder what was going on.

Thank you :pray:

hanna-paasivirta commented:
In the latest commits I have:

- Changed sys arg in entry.py – I think there was an index typo that meant the output json path would not be read
- Changed the main function in embeddings_demo.py to reflect what entry.py expects (json input and output)
- Renamed from generate_embedings to embeddings for clarity and to match entry.py requirements
- Renamed create_collection to add_docs
- Removed the main function in embeddings.py as it will not normally be used directly from this folder
- Added READMEs
- not included the demo.py from PR #121 , as I think it does the same thing as embeddings_demo.py now.

hanna-paasivirta commented:
Got it, does it look right now? I moved the vector store set up and instance creation code inside embeddings/demo/example.py

josephjclark commented on services/embeddings/embeddings.py:
```diff
@@ -0,0 +1,109 @@
+    
+    def add_docs(self, docs, drop_old=True, **kwargs):
+        """
+        Create a new vectorstore from documents and initialise it with the specified settings.
```
Is this comment still accurate? Looks like it's fallen out of date to me - I don't think this creates a new vector store at all :thinking:

hanna-paasivirta replied:
The LangChain function from_documents will create a new collection while adding documents. I'll edit the comment from "vectorstore" to "collection" to clarify.

josephjclark commented on services/embeddings_demo/embeddings_demo.py:
```diff
@@ -0,0 +1,37 @@
+import os
+from dotenv import load_dotenv
+from embeddings.demo.example import get_demo_vectorstore
```
I'd really like a cleaner import here :thinking: 

How can we set up the package structure to support this?
```
from embeddings import demo

demo.add_docs()
demo.search()
```

That would be really neat. I was thinking about having a sub folder per file, which allows stores to have many supporting files and still give us a tidy folder structure. If that's in our way, maybe we have a top level entry point file (`example.py`), AND a subfolder to hide  away the messy stuff, if needed (`example/setup.py`, `example/data.json`)

We'd also have to do away with the `get_demo_vectorstore` and initialise the store when the module is loaded. Which, as I've asked elsewhere, I'm  not sure what the cost of that is.

If we really have to have an initialisation function, maybe we can just call it `connect`? Which I expect to give me a unique connection to the underlying database.

That would give me code like:

```
from embeddings import demo

store = demo.connect()
store.add_docs(...)
store.search(...)
```

hanna-paasivirta replied:
It seems a little unexpected to me to do something potentially costly/unwanted (wiping a database) in the background on just an import. If it's ok, I'll just rename it to connect() to make it clear that something will happen when you use the class instance.

josephjclark replied:
Yeah agreed!

`connect` should just give me a live connection to a database, which I can then do as I want. And it's fairly clear from that semantics that that's what I expect. So that's cool!

One note on this: in the example store, the `connect` function will also load data. Which is a little funky because if two  different services connect, they'll duplicate the data load.

Maybe just for the purposes of the demo, we can have a `connect()` and a `load_default_data()` function in the example store, and then the demo script will explicitly connect AND load the data?

This is also a  bit weird - we'd usually have some offline process to populate the store and not make it part of apollo like this. But it keeps the data loading logic in the store (which I like) and takes side-effects out of `connect` (which I like). So I think it's an improvement?

We'll have to look more closely at when and how the embedding stories get loaded. But we'll defer that here - we're going to bump into it pretty quickly on your mapper!

hanna-paasivirta replied:
I added a load_default_data() in the demo – does it look like what you meant?

josephjclark replied:
Perfect!

josephjclark reviewed (commented):
Hey @hanna-paasivirta 

This is close, thank you!

I find it a little weird that the `demo` store has an `example.py`. I'd like to just use the word `example` or `demo` - and I think I prefer example because it's an example data set which is _used_ by the demo.  But this mixing of language is hard on my brain and I keep getting mixed up on what I'm looking at :exploding_head: 

The other big thing is that the data loading still isn't super clear to me, and I'd like to see a little bit of tidying up around there.

The idea goal of this PR is a really clear example and just enough documentation that I could ask someone like Hunter or Elias to add a new embedding store, or to port the existing docs embedding into this structure. I don't think we're quite there yet.

josephjclark commented on services/entry.py:
```diff
@@ -37,7 +37,7 @@ def call(service, input_path, output_path):
     output_path = None
 
-    if len(sys.argv) == 5:
+    if len(sys.argv) == 4:
```
TODO for me: remember to test this with CLI

hanna-paasivirta commented:
I have just pushed a commit where I did the following:

- Moved example.py out of the demo folder onto the same level as embeddings.py
- Renamed get_demo_vectorstore to connect
- Changed to import whole module instead of specific functions
- Edited the comments you noted
- Added the demo and an explanation of its structure in the README of embeddings

Did I miss or misunderstand anything?

josephjclark reviewed (approved):
Thank you @hanna-paasivirta ! :tada: 

Just one last thing from you please (sorry I should have mentioned this earlier)

Can you run `bun changeset`  and follow the prompt to generate  a changeset file with a vague user-facing summary of what this PR does. Something as simple as  `Add an embedding service` will suffice for now.

This should be a minor change, because we're adding something new.

Then commit and push the file to here,and I can merge.

Happy to do this live on the call with you tomorrow morning :)

josephjclark commented on services/embeddings/example.py:
```diff
@@ -0,0 +1,32 @@
+        vectorstore_type="zilliz",
+        embedding_type="openai",
+        connection_args = {
+            "uri": os.getenv('ZILLIZ_CLOUD_URI'),
```
Oh @hanna-paasivirta please make sure this uses the same env var as the basic search service :pray:
